File: incre-tests/synduce/numbers/int_nat_toint.f
Inductive Nat = z Unit | s Nat;
Inductive INat = positive Nat | negative Nat;

nsum = fix (
  \f: Nat -> Int. \n: Nat.
  match n with
    z _ -> 0
  | s m -> + 1 (f m)
  end
);

itoint = \n: INat.
  match n with
    negative m -> - 0 (+ 1 (nsum m))
  | positive m -> nsum m
  end;

Inductive EInt = zero Unit | sub1 EInt | add1 EInt;

repr = fix (
  \f: EInt -> INat. \e: EInt.
  match e with
    zero _ -> positive (z unit)
  | sub1 e' ->
    let res = f e' in
    match res with
      positive (z _) -> negative (z unit)
    | positive (s n) -> positive n
    | negative n -> negative (s n)
    end
  | add1 e' ->
    let res = f e' in
    match res with
      positive n -> positive (s n)
    | negative (z _) -> positive (z unit)
    | negative (s n) -> negative n
    end
  end
);

target = fix (
  \f: EInt -> Compress EInt. \e: EInt.
  match e with
    zero _ -> zero unit
  | add1 e' -> add1 (f e')
  | sub1 e' -> sub1 (f e')
  end
);

main = \x: EInt. itoint (repr (target x));

File: ml-benchmark/synduce/numbers/int_nat_toint.ml
type nat = Z of unit | S of nat
type inat = Positive of nat | Negative of nat
type eint = Zero of unit | Sub1 of eint | Add1 of eint

let rec nsum n =
  match n with
  | Z _ -> 0
  | S m -> 1 + nsum m

let itoint n =
  match n with
  | Negative m -> -(1 + nsum m)
  | Positive m -> nsum m

let rec repr e =
  match e with
  | Zero _ -> Positive (Z ())
  | Sub1 e' ->
      let res = repr e' in
      match res with
      | Positive n ->
          (match n with
           | Z _ -> Negative (Z ())
           | S n' -> Positive n')
      | Negative n -> Negative (S n)
  | Add1 e' ->
      let res = repr e' in
      match res with
      | Positive n -> Positive (S n)
      | Negative n ->
          (match n with
           | Z _ -> Positive (Z ())
           | S n' -> Negative n')

let rec target e =
  match e with
  | Zero u -> Zero u
  | Add1 e' -> Add1 (target e')
  | Sub1 e' -> Sub1 (target e')

let program x = itoint (repr (target x))